### PR TITLE
jscookie: small changes to make it more correct. Might break existing code?

### PR DESCRIPTION
**Pull request by webbiesdk**
These changes will make the js-cookie declaration "more correct". 
However, the changes might break existing code, and I'm therefore not sure if they should be included.

The changes:

- `noConflict` marked as being maybe present: To reflect that the method is not exposed on AMD or CommonJS environments. 
- Changed the type of `value` in the `write` method from `string` to `string | object`. 
The `value` argument will be an object if `JSON.stringify` throws an exception while trying to serialize the value given in the `set` method. Here is an example: 
```JavaScript
var foo = {};
foo.foo = foo;
Cookies.withConverter({
	read: function (value) {return value;}, 
	write: function(value, name){console.log(typeof value); return "foo";}
}).set("foo", foo)
```
The above code will print `object` into the console.   

I could also have changed the type of `const Cookies` to `Cookies.CookiesStatic | undefined` (since the `noConflict`, method sets `Cookies` to undefined). 
However, this would surely break every existing usage of js-cookie, so I've not included that change. 

Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/js-cookie/js-cookie-tests.ts [hunk 1]
```diff
@@ -19,8 +19,7 @@ Cookies.get();
 Cookies.remove('name');
 Cookies.remove('name', { path: '' });
 
-const Cookies2 = Cookies.noConflict();
-Cookies2; // $ExpectType CookiesStatic
+const Cookies2: Cookies.CookiesStatic = (Cookies.noConflict as () => Cookies.CookiesStatic)();
```
Comment (BendingBender) [inline types/js-cookie/js-cookie-tests.ts]: It was much clearer before this change, why did you change it?

File: types/js-cookie/js-cookie-tests.ts [hunk 2]
```diff
@@ -41,11 +40,11 @@ Cookies.defaults.path = '';
 delete Cookies.defaults.path;
 
 const PHPCookies = Cookies.withConverter({
-    write(value) {
-        return encodeURIComponent(value)
+    write(value: string | object) {
```
Comment (BendingBender) [inline types/js-cookie/js-cookie-tests.ts]: Not a big thing, but it's actually better to check it like this:
```typescript
write(value) {
  value; // $ExpectType string | object
  // ...
}
```
The reasoning is that this way you verify that the interface doesn't do something like `write(value: any) => void`, which you would miss with just a type signatute for the `write` method.

File: types/js-cookie/js-cookie-tests.ts [hunk 3]
```diff
@@ -41,11 +40,11 @@ Cookies.defaults.path = '';
...
+        return encodeURIComponent(value as string)
             .replace(/%(23|24|26|3A|3C|3E|3D|2F|3F|40|5B|5D|5E|60|7B|7D|7C)/g, decodeURIComponent);
     },
-    read(value) {
+    read(value: string) {
```
Comment (BendingBender) [inline types/js-cookie/js-cookie-tests.ts]: same as before

File: types/js-cookie/index.d.ts [hunk 4]
```diff
@@ -79,7 +79,7 @@ declare namespace Cookies {
          * or SDK. Note: The noConflict method is not necessary when using
          * AMD or CommonJS, thus it is not exposed in those environments.
          */
-        noConflict(): CookiesStatic;
+        noConflict?(): CookiesStatic;
```
Comment (BendingBender) [inline types/js-cookie/index.d.ts]: Is this method really optional?

File: types/js-cookie/index.d.ts [hunk 5]
```diff
@@ -88,10 +88,11 @@ declare namespace Cookies {
          * will run the converter first for each cookie. The returned
          * string will be used as the cookie value.
          */
-        withConverter(converter: CookieConverter | { write: CookieConverter; read: CookieConverter; }): CookiesStatic;
+        withConverter(converter: CookieReadConverter | { write: CookieWriteConverter; read: CookieReadConverter; }): CookiesStatic;
     }
 
-    type CookieConverter = (value: string, name: string) => string;
+    type CookieWriteConverter = (value: string | object, name: string) => string;
```
Comment (BendingBender) [inline types/js-cookie/index.d.ts]: You shouldn't use `object` here. If you do, users always will have to convert this to an object with an index signature: `{ [key: string]: any; }` to access properties (or something more specific). It is better if you type it as an object with a type signature directly to spare your users this unnecessary conversion. And I suspect that the values aren't really `any` typed, I suppose they are `string`s or `string[]` or something like this.

**Review by BendingBender — CHANGES_REQUESTED** [review]

File: types/js-cookie/index.d.ts (hunk 5 above)
Comment (webbiesdk) [reply types/js-cookie/index.d.ts]: I used `object` since this reflected the type used in the `set` method. 

Since the value is passed directly from the argument in the `set`method to the `CookieWriteConverter`, I felt the two should have the same type.

File: types/js-cookie/js-cookie-tests.ts (hunk 1 above)
Comment (webbiesdk) [reply types/js-cookie/js-cookie-tests.ts]: Are you talking about that I removed the `$ExpectType`? Or that the noConflict method is now optional? 

I'll reintroduce the `$ExpectType`, thanks for explaining what it does.

File: types/js-cookie/index.d.ts (hunk 4 above)
Comment (webbiesdk) [reply types/js-cookie/index.d.ts]: As stated in the line above: 
`The noConflict method is not necessary when using AMD or CommonJS, thus it is not exposed in those environments.`
The `noConflict` method only exists in a browser environment and is thus optional.

File: types/js-cookie/index.d.ts (hunk 5 above)
Comment (BendingBender) [reply types/js-cookie/index.d.ts]: Ah, OK. Then you should do the following:
```typescript
interface CookiesStatic<T extends object = object> {
  // ...
  set(name: string, value: string | T, options?: CookieAttributes): void;
  // ...
  withConverter(converter: CookieReadConverter | { write: CookieWriteConverter<T>; read: CookieReadConverter; }): CookiesStatic;
}

type CookieWriteConverter<T extends object> = (value: string | T, name: string) => string;
```

File: types/js-cookie/index.d.ts (hunk 4 above)
Comment (BendingBender) [reply types/js-cookie/index.d.ts]: Fair enough. Missed this one.

File: types/js-cookie/js-cookie-tests.ts (hunk 1 above)
Comment (BendingBender) [reply types/js-cookie/js-cookie-tests.ts]: You can do the call like this, without explicit cast:
```typescript
const Cookies2 = Cookies!.noConflict();
```

Comment (webbiesdk) [reply types/js-cookie/js-cookie-tests.ts]: Thanks. I saw the bang operator back when it was introduced, loved it, then forgot about it. 

Btw. its like this:  `const Cookies2 = Cookies.noConflict!();`

Comment (BendingBender) [reply types/js-cookie/js-cookie-tests.ts]: Oh, thanks!

File: types/js-cookie/index.d.ts (hunk 5 above)
Comment (BendingBender) [reply types/js-cookie/index.d.ts]: Sorry, my original version was incorrect. Typed it down without testing.

Comment (BendingBender) [reply types/js-cookie/index.d.ts]:
```typescript
interface CookiesStatic<T extends object = object> {
  // ...
  set(name: string, value: string | T, options?: CookieAttributes): void;
  // ...
  withConverter<TConv extends object>(
    converter:
      | CookieReadConverter
      | { write: CookieWriteConverter<TConv>; read: CookieReadConverter }
  ): CookiesStatic<TConv>;
}

type CookieWriteConverter<T extends object> = (value: string | T, name: string) => string;
```

**Review by BendingBender — APPROVED** [review]
